### PR TITLE
scr: require dtcmp, switch from /tmp to /dev/shm

### PR DESCRIPTION
**Pull request by adammoody**
I have two questions regarding this PR.

Newer versions of SCR now *require* DTCMP as a dependency rather than allowing it as an option.  It would be nice to disable the ``+dtcmp`` variant for those later SCR versions, but I don't see a way to do that.

Is there a way?

I did find this related issue:
https://github.com/spack/spack/issues/9740

Also, I'd like to change a particular directory from ``/tmp`` to ``/dev/shm``.  While this is a better choice on Linux systems, ``/dev/shm`` does not exist on other platforms like macOS.  It'd be nice to use ``/dev/shm`` on Linux but keep ``/tmp`` on other platforms.

Is there a recommended way to set the variant default based on the target platform?

Thanks for any help!

### CODE REVIEW

File: var/spack/repos/builtin/packages/scr/package.py [hunk 1]
```diff
@@ -83,9 +85,9 @@ class Scr(CMakePackage):
             multi=False,
             description='File locking style for SCR.')
 
-    variant('cache_base', default='/tmp',
+    variant('cache_base', default='/dev/shm',
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/scr/package.py:88]: Curious.  Is `/dev/shm` a directory that is expected to exist by default or is it created by the internal build process of the package?

**Comment by tldahlgren** [comment]
> I have two questions regarding this PR.

Great questions.  Pulling @becker33 and @scheibelp in in case they have ideas.

> Also, I'd like to change a particular directory from `/tmp` to `/dev/shm`. While this is a better choice on Linux systems, `/dev/shm` does not exist on other platforms like macOS. It'd be nice to use `/dev/shm` on Linux but keep `/tmp` on other platforms.
> 
> Is there a recommended way to set the variant default based on the target platform?

Clearly changing to `/dev/shm` is not portable so IMO should not be done.  The only thing I can think of off the top of my head is to possibly use a neutral default, such as an empty string (``), that is then substituted -- perhaps using a property -- with the target platform-specific directory.

**Comment by gonsie** [comment]
Your two dtcmp depends_on statements look good. As mentioned in #9740, we can use an additional conflicts statement to ensure users don’t try to turn off DTCMP:

```
conflicts('~dtcmp', when="@3:", msg="<SCR> DTCMP required for versions >=3")
```

I would use this for the temp directory default (thanks @CamStan) (this requires an `import sys` at the top):

```
platform_tmp_default = '/dev/shm' if sys.platform=='linux' else '/tmp'
variant('cache_base', default=platform_tmp_default,
        description='Compile time default location for checkpoint cache.'
        )
variant('cntl_base', default=platform_tmp_default,
        description='Compile time default location for control directory.')
```

**Comment by adammoody** [comment]
Thanks for the feedback everyone!

Thanks @gonsie and @CamStan for the improvements.  I pushed those as a second commit which I can squash later.

@tldahlgren , yes, I agree that it's not ideal to pick a value that is not portable.  We'd like to avoid using ``/tmp`` as a universal default, since that path can impose a performance hit to Linux users.  How about this conditional setting?

**Review by gonsie — DISMISSED** [review]
Just tested on a mac and linux system. Works great!

**Review by tldahlgren — APPROVED** [review]
LGTM.  Thanks for the changes.

**Comment by tldahlgren** [comment]
Thank you for your contribution!

**Comment by adammoody** [comment]
Thanks for your time and help, @tldahlgren

**Pull request merged**